[FIX] No Invalid Char is allowed in request

sarthakpranesh opened the pull request:
1. updated submit() to check for any invalid character in the username
2. if found, the request to update profile is not made and an alert is displayed

***Please verify, but the Server Version: 3.0.0-develop does not allow the username to have dot.

@RocketChat/ReactNative
Closes #1682

djorkaeffalexandre commented on app/views/ProfileView/index.js:
```diff
@@ -180,7 +185,26 @@ class ProfileView extends React.Component {
 		if (user.username !== username) {
-			params.username = username;
+			if (!username.match(/[$&+,:;=?@#|'.<>^* ()%!]/g)) {
+				params.username = username;
```
You can't use this regex here, because on our servers we have a configuration about regex usernames and channel names, you need to get this setting from server and try to match.
To find this setting on your own server go to `Administration -> General -> UTF8`.
![Screen Shot 2020-02-08 at 14 48 30](https://user-images.githubusercontent.com/29778115/74089641-2d092780-4a82-11ea-8533-7a20bd55c6b3.png)

sarthakpranesh replied:
Ok cool, thanks for letting me know about this. I will make the required changes.
:+1:

sarthakpranesh commented:
1. Used server settings for username validation
2. Added necessary text to i18n and updated ProfileView alert to use it

djorkaeffalexandre commented on app/i18n/locales/en.js:
```diff
@@ -496,5 +496,8 @@ export default {
-	This_will_clear_all_your_offline_data: 'This will clear all your offline data.'
+	This_will_clear_all_your_offline_data: 'This will clear all your offline data.',
+	Incorrect_username: 'Incorrect Username',
+	Username_can_only_have_letters_numbers_dots_hyphens_and_underscores: 'Username can only have letters, numbers, dots, hyphens and underscores',
```
We can't always confirm that, I think you can use just: `Username has chars not allowed by server.`

sarthakpranesh replied:
Ok, yeah sure

djorkaeffalexandre commented on app/i18n/locales/en.js:
```diff
@@ -496,5 +496,8 @@ export default {
+	This_will_clear_all_your_offline_data: 'This will clear all your offline data.',
+	Incorrect_username: 'Incorrect Username',
+	Username_can_only_have_letters_numbers_dots_hyphens_and_underscores: 'Username can only have letters, numbers, dots, hyphens and underscores',
+	Okay: 'Okay'
```
If you follow the bellow comment you can remove this.

djorkaeffalexandre commented on app/views/ProfileView/index.js:
```diff
@@ -1,6 +1,11 @@
+	View,
+	ScrollView,
+	Keyboard,
+	Alert
```
Remember to remove these changes when you switch from `Alert` to `showErrorAlert`.

djorkaeffalexandre commented on app/views/ProfileView/index.js:
```diff
@@ -181,7 +187,27 @@ class ProfileView extends React.Component {
 		if (user.username !== username) {
-			params.username = username;
+			const regEx = new RegExp(`^${ UTF8_Names_Validation }$`);
+			if (regEx.test(username)) {
```
instead of make this logic here, I think it's better if you create a function like: `validUsername`.
```javascript
validUsername = () => {
  const { username } = this.state;
  const { UTF8_Names_Validation } = this.props;
  const usernameRegex = new RegExp(`^${ UTF8_Names_Validation }$`);
  if (usernameRegex.test(username)) {
    return true;
  }
  showErrorAlert(yourMessage, yourTitle);
  return false;
}
```
so, you can use only: 
```javascript
if (this.validUsername()) {
  params.username = username;
} else {
  this.setState({ saving: false });
  return;
}
```

sarthakpranesh replied:
Yeah I agree on that, it'll be cleaner than

djorkaeffalexandre commented on app/views/ProfileView/index.js:
```diff
@@ -181,7 +187,27 @@ class ProfileView extends React.Component {
+					saving: false,
+					currentPassword: null
+				});
+				Alert.alert(
```
Use `showErrorAlert` instead `Alert` to show error alerts.

sarthakpranesh replied:
Ok sure

sarthakpranesh commented:
Updates
1. Removed Alert from import and reverted changed
2. Updated the message text `Username has chars not allowed by server`
3. Cleaned the code and made another function for username validation

sarthakpranesh commented:
FIXED
All three RegistrationView, SetUsernameView and ProfileView check for the entered `username` to make sure it matches the settings on the server.

Please let me know of any further changes :smile: 
![RegistrationView](https://user-images.githubusercontent.com/41206172/75905535-0664c380-5e6b-11ea-8d0f-e1bb44251591.gif)
![profileView](https://user-images.githubusercontent.com/41206172/75905545-0b297780-5e6b-11ea-8a9b-b32749c0af1c.gif)
